source/defender_demo.c

```diff
@@ -0,0 +1,945 @@
+ * false otherwise.
+ */
+static bool prvCollectDeviceMetrics( void );
+
+/**
+ * @brief Generate the device defender report.
+ *
+ * @param[out] pulOutReportLength Length of the device defender report.
+ *
+ * @return true if the report is generated successfully;
+ * false otherwise.
+ */
+static bool prvGenerateDeviceMetricsReport( uint32_t * pulOutReportLength );
+
+/**
+ * @brief Publish the generated device defender report.
+ *
+ * @param[in] ulReportLength Length of the device defender report.
+ *
+ * @return true if the report is published successfully;
+ * false otherwise.
+ */
+static bool prvPublishDeviceMetricsReport( uint32_t ulReportLength );
+
+/**
+ * @brief Validate the response received from the AWS IoT Device Defender Service.
+ *
+ * This functions checks that a valid JSON is received and the report ID
+ * is same as was sent in the published report.
+ *
+ * @param[in] pcDefenderResponse The defender response to validate.
+ * @param[in] ulDefenderResponseLength Length of the defender response.
+ *
+ * @return true if the response is valid;
+ * false otherwise.
+ */
+static bool prvValidateDefenderResponse( const char * pcDefenderResponse,
+                                         uint32_t ulDefenderResponseLength );
+
+/**
+ * @brief The task used to demonstrate the Defender API.
+ *
+ * This task collects metrics from the device using the functions in
+ * metrics_collector.h and uses them to build a defender report using functions
+ * in report_builder.h. Metrics include the number for bytes written and read
+ * over the network, open TCP and UDP ports, and open TCP sockets. The
+ * generated report is then published to the AWS IoT Device Defender service.
+ *
+ * @param[in] pvParameters Parameters as passed at the time of task creation.
+ * Not used in this example.
+ */
+static void prvDefenderDemoTask( void * pvParameters );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Create the task that demonstrates the Device Defender library API.
+ */
+void vStartDefenderDemo( configSTACK_DEPTH_TYPE uxStackSize,
+                         UBaseType_t uxPriority )
+{
+    xTaskCreate( prvDefenderDemoTask, /* Function that implements the task. */
+                 "Defender",          /* Text name for the task - only used for debugging. */
+                 uxStackSize,         /* Size of stack (in words, not bytes) to allocate for the task. */
+                 NULL,                /* Task parameter - not used in this case. */
+                 uxPriority,          /* Task priority, must be between 0 and configMAX_PRIORITIES - 1. */
+                 NULL );              /* Used to pass out a handle to the created task - not used in this case. */
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvSubscribeToDefenderTopics( void )
+{
+    MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    MQTTSubscribeInfo_t xSubscribeInfo[ 2 ];
```

Comment:
These structs need to be static since they must persist until the command is processed